File: src-tauri/src/scraper/parser.rs
// TikTok Parser Module
// HTML/JSON parsing for TikTok Shop pages

use anyhow::{Context, Result};
use chromiumoxide::Page;
use scraper::{Html, Selector};
use serde_json::Value;
use uuid::Uuid;

use crate::models::Product;

pub struct TikTokParser {
    selectors: Vec<String>,
}

impl TikTokParser {
    pub fn new(selectors: Option<Vec<String>>) -> Self {
        Self {
            selectors: selectors.unwrap_or_else(|| {
                vec![
                    "[data-e2e='product-card']".to_string(),
                    ".product-card".to_string(),
                    ".product-item".to_string(),
                ]
            }),
        }
    }

    pub async fn parse_product_list(&self, page: &Page) -> Result<Vec<Product>> {
        // Try JavaScript first (faster and more reliable)
        log::debug!("Attempting to parse products from __INITIAL_STATE__");

        let script = r#"
            (() => {
                if (window.__INITIAL_STATE__) {
                    // Direct product list
                    if (window.__INITIAL_STATE__.products) {
                        return JSON.stringify(window.__INITIAL_STATE__.products);
                    }
                    // Product list wrapper
                    if (window.__INITIAL_STATE__.productList && window.__INITIAL_STATE__.productList.products) {
                        return JSON.stringify(window.__INITIAL_STATE__.productList.products);
                    }
                    // Shop page structure
                    if (window.__INITIAL_STATE__.shop && window.__INITIAL_STATE__.shop.products) {
                        return JSON.stringify(window.__INITIAL_STATE__.shop.products);
                    }
                    // General search result
                    if (window.__INITIAL_STATE__.search && window.__INITIAL_STATE__.search.item_list) {
                        return JSON.stringify(window.__INITIAL_STATE__.search.item_list);
                    }
                }
                
                // Try to find JSON in script tags (SIGI_STATE is common in TikTok)
                const sigiState = document.getElementById('SIGI_STATE');
                if (sigiState) {
                    try {
                        const data = JSON.parse(sigiState.textContent);
                        if (data.ItemModule) {
                            return JSON.stringify(Object.values(data.ItemModule));
                        }
                    } catch (e) {}
                }
                
                return null;
            })()
        "#;

        let result = page.evaluate(script).await?;

        if let Some(json_str) = result.value() {
            if !json_str.is_null() {
                if let Ok(json_text) = serde_json::from_value::<String>(json_str.clone()) {
                    if let Ok(products_json) = serde_json::from_str::<Value>(&json_text) {
                        if let Some(arr) = products_json.as_array() {
                            let products: Vec<Product> = arr
                                .iter()
                                .filter_map(|item| self.parse_product_json(item).ok())
                                .collect();

                            if !products.is_empty() {
                                log::info!("Parsed {} products from JSON", products.len());
                                return Ok(products);
                            }
                        }
                    }
                }
            }
        }

        // Fallback to DOM parsing
        log::debug!("Falling back to DOM parsing");
        self.parse_product_list_from_dom(page).await
    }

    async fn parse_product_list_from_dom(&self, page: &Page) -> Result<Vec<Product>> {
        let html = page.content().await?;
        let document = Html::parse_document(&html);

        for selector_str in &self.selectors {
            if let Ok(selector) = Selector::parse(selector_str) {
                let elements: Vec<_> = document.select(&selector).collect();

                if !elements.is_empty() {
                    log::debug!(
                        "Found {} products with selector: {}",
                        elements.len(),
                        selector_str
                    );

                    let products: Vec<Product> = elements
                        .iter()
                        .filter_map(|element| self.parse_product_element(element).ok())
                        .collect();

                    if !products.is_empty() {
                        return Ok(products);
                    }
                }
            }
        }

        log::warn!("No products found in DOM");
        Ok(Vec::new())
    }

    fn parse_product_json(&self, data: &Value) -> Result<Product> {
        let tiktok_id = data
            .get("id")
            .or_else(|| data.get("productId"))
            .or_else(|| data.get("product_id"))
            .and_then(|v| {
                v.as_str().or_else(|| {
                    v.as_i64()
                        .map(|n| Box::leak(n.to_string().into_boxed_str()) as &str)
                })
            })
            .context("Missing product ID")?
            .to_string();

        let title = data
            .get("title")
            .or_else(|| data.get("name"))
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();

        let price = self.extract_price(data.get("price"))?;
        let original_price_val = self
            .extract_price(
                data.get("originalPrice")
                    .or_else(|| data.get("original_price")),
            )
            .ok();

        let currency = data
            .get("currency")
            .and_then(|v| v.as_str())
            .unwrap_or("BRL")
            .to_string();

        Ok(Product {
            id: Uuid::new_v4().to_string(),
            tiktok_id: tiktok_id.clone(),
            title,
            description: data
                .get("description")
                .and_then(|v| v.as_str())
                .map(String::from),
            price,
            original_price: original_price_val,
            currency,
            category: data
                .get("category")
                .and_then(|v| v.as_str())
                .map(String::from),
            subcategory: data
                .get("subcategory")
                .and_then(|v| v.as_str())
                .map(String::from),
            seller_name: data
                .get("seller")
                .and_then(|v| v.get("name"))
                .and_then(|v| v.as_str())
                .map(String::from),
            seller_rating: data
                .get("seller")
                .and_then(|v| v.get("rating"))
                .and_then(|v| v.as_f64()),
            product_rating: data.get("rating").and_then(|v| Self::extract_rating(v)),
            reviews_count: data
                .get("reviewCount")
                .and_then(|v| v.as_i64())
                .unwrap_or(0) as i32,
            sales_count: self.parse_sales_count(data.get("salesCount"))?,
            sales_7d: self.parse_sales_count(data.get("sales7d"))?,
            sales_30d: self.parse_sales_count(data.get("sales30d"))?,
            commission_rate: data.get("commissionRate").and_then(|v| v.as_f64()),
            image_url: data
                .get("imageUrl")
                .or_else(|| data.get("image"))
                .and_then(|v| v.as_str())
                .map(String::from),
            images: data
                .get("images")
                .and_then(|v| v.as_array())
                .map(|arr| {
                    arr.iter()
                        .filter_map(|v| v.as_str().map(String::from))
                        .collect()
                })
                .unwrap_or_default(),
            video_url: data
                .get("videoUrl")
                .and_then(|v| v.as_str())
                .map(String::from),
            product_url: data
                .get("url")
                .and_then(|v| v.as_str())
                .map(String::from)
                .unwrap_or_else(|| format!("https://shop.tiktok.com/product/{}", &tiktok_id)),
            affiliate_url: data
                .get("affiliateUrl")
                .and_then(|v| v.as_str())
                .map(String::from),
            has_free_shipping: data
                .get("freeShipping")
                .and_then(|v| v.as_bool())
                .unwrap_or(false),
            is_trending: data
                .get("isTrending")
                .and_then(|v| v.as_bool())
                .unwrap_or(false),
            is_on_sale: original_price_val.map_or(false, |op| op > price),
            in_stock: data
                .get("inStock")
                .and_then(|v| v.as_bool())
                .unwrap_or(true),
            stock_level: data
                .get("stock")
                .or_else(|| data.get("stockLevel"))
                .or_else(|| data.get("quantity"))
                .and_then(|v| v.as_i64())
                .map(|v| v as i32),
            collected_at: chrono::Utc::now().to_rfc3339(),
            updated_at: chrono::Utc::now().to_rfc3339(),
        })
    }

    fn parse_product_element(&self, element: &scraper::ElementRef) -> Result<Product> {
        let title_selector =
            Selector::parse("[data-e2e='product-title'], .product-title, h3, h4").ok();
        let title = if let Some(sel) = title_selector {
            element
                .select(&sel)
                .next()
                .map(|e| e.text().collect::<String>().trim().to_string())
                .unwrap_or_default()
        } else {
            String::new()
        };

        let price_selector =
            Selector::parse("[data-e2e='product-price'], .product-price, .price").ok();
        let price_text = if let Some(sel) = price_selector {
            element
                .select(&sel)
                .next()
                .map(|e| e.text().collect::<String>())
                .unwrap_or_else(|| "0".to_string())
        } else {
            "0".to_string()
        };
        let price = Self::parse_price_text(&price_text);

        let image_selector = Selector::parse("img").ok();
        let image_url = if let Some(sel) = image_selector {
            element
                .select(&sel)
                .next()
                .and_then(|e| e.value().attr("src"))
                .map(String::from)
        } else {
            None
        };

        let link_selector = Selector::parse("a").ok();
        let product_url = if let Some(sel) = link_selector {
            element
                .select(&sel)
                .next()
                .and_then(|e| e.value().attr("href"))
                .map(String::from)
                .unwrap_or_default()
        } else {
            String::new()
        };

        let tiktok_id =
            Self::extract_id_from_url(&product_url).unwrap_or_else(|| Uuid::new_v4().to_string());

        Ok(Product {
            id: Uuid::new_v4().to_string(),
            tiktok_id,
            title,
            description: None,
            price,
            original_price: None,
            currency: "BRL".to_string(),
            category: None,
            subcategory: None,
            seller_name: None,
            seller_rating: None,
            product_rating: None,
            reviews_count: 0,
            sales_count: 0,
            sales_7d: 0,
            sales_30d: 0,
            commission_rate: None,
            image_url,
            images: vec![],
            video_url: None,
            product_url,
            affiliate_url: None,
            has_free_shipping: false,
            is_trending: false,
            is_on_sale: false,
            in_stock: true,
            stock_level: None,
            collected_at: chrono::Utc::now().to_rfc3339(),
            updated_at: chrono::Utc::now().to_rfc3339(),
        })
    }

    fn extract_price(&self, value: Option<&Value>) -> Result<f64> {
        if let Some(v) = value {
            if let Some(num) = v.as_f64() {
                return Ok(num);
            }
            if let Some(s) = v.as_str() {
                return Ok(Self::parse_price_text(s));
            }
            if let Some(obj) = v.as_object() {
                if let Some(val) = obj.get("value").and_then(|v| v.as_f64()) {
                    return Ok(val);
                }
            }
        }
        Ok(0.0)
    }

    fn parse_price_text(text: &str) -> f64 {
        // Keep only digits, comma, dot
        let cleaned: String = text
            .chars()
            .filter(|c| c.is_ascii_digit() || *c == ',' || *c == '.')
            .collect();

        if cleaned.is_empty() {
            return 0.0;
        }

        // Check for multiple separators to determine format
        let last_comma = cleaned.rfind(',');
        let last_dot = cleaned.rfind('.');

        let normalized = match (last_comma, last_dot) {
            (Some(c), Some(d)) => {
                if c > d {
                    // Format: 1.234,56 (BR/EU) -> Remove dots, replace comma with dot
                    cleaned.replace('.', "").replace(',', ".")
                } else {
                    // Format: 1,234.56 (US) -> Remove commas
                    cleaned.replace(',', "")
                }
            }
            (Some(_), None) => {
                // Format: 1234,56 (BR/EU) -> Replace comma with dot
                cleaned.replace(',', ".")
            }
            (None, Some(d)) => {
                // Format: 1234.56 or 1.234.567
                let dot_count = cleaned.matches('.').count();
                if dot_count > 1 {
                    // Multiple dots = thousands separators (1.234.567)
                    cleaned.replace('.', "")
                } else if cleaned.len() - d - 1 == 3 {
                    // One dot, 3 digits after = likely thousands (1.234)
                    // This is a heuristic, but safe for TikTok BR context
                    cleaned.replace('.', "")
                } else {
                    // One dot, not 3 digits = decimal (1.23)
                    cleaned
                }
            }
            (None, None) => cleaned,
        };

        normalized.parse().unwrap_or(0.0)
    }

    fn parse_sales_count(&self, value: Option<&Value>) -> Result<i32> {
        if let Some(v) = value {
            if let Some(num) = v.as_i64() {
                return Ok(num as i32);
            }
            if let Some(s) = v.as_str() {
                return Ok(Self::parse_sales_text(s));
            }
        }
        Ok(0)
    }

    fn parse_sales_text(text: &str) -> i32 {
        let text_lower = text.to_lowercase();

        // Check for suffixes
        let multiplier = if text_lower.contains('k') {
            1000.0
        } else if text_lower.contains('m') {
            1000000.0
        } else {
            1.0
        };

        if multiplier > 1.0 {
            // Handle 1.5k or 1,5k
            let num_part = text_lower
                .trim_end_matches('k')
                .trim_end_matches('m')
                .trim()
                .replace(',', "."); // Normalize decimal separator

            if let Ok(val) = num_part.parse::<f64>() {
                return (val * multiplier) as i32;
            }
        }

        // Fallback: extract all digits (handles 1.234 as 1234)
        let digits: String = text.chars().filter(|c| c.is_ascii_digit()).collect();
        digits.parse().unwrap_or(0)
    }

    fn extract_rating(value: &Value) -> Option<f64> {
        if let Some(num) = value.as_f64() {
            if num >= 0.0 && num <= 5.0 {
                return Some(num);
            }
        }
        if let Some(obj) = value.as_object() {
            if let Some(avg) = obj.get("average").and_then(|v| v.as_f64()) {
                return Some(avg);
            }
        }
        None
    }

    fn extract_id_from_url(url: &str) -> Option<String> {
        use regex::Regex;
        let re = Regex::new(r"/product/(\d+)").ok()?;
        re.captures(url)
            .and_then(|cap| cap.get(1))
            .map(|m| m.as_str().to_string())
    }
}

impl Default for TikTokParser {
    fn default() -> Self {
        Self::new(None)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_parse_stock_level() {
        let parser = TikTokParser::default();

        // Case 1: "stock" field
        let data = json!({
            "id": "123",
            "title": "Test Product",
            "price": 10.0,
            "stock": 100
        });
        let product = parser.parse_product_json(&data).unwrap();
        assert_eq!(product.stock_level, Some(100));

        // Case 2: "stockLevel" field
        let data = json!({
            "id": "124",
            "title": "Test Product 2",
            "price": 10.0,
            "stockLevel": 50
        });
        let product = parser.parse_product_json(&data).unwrap();
        assert_eq!(product.stock_level, Some(50));

        // Case 3: "quantity" field
        let data = json!({
            "id": "125",
            "title": "Test Product 3",
            "price": 10.0,
            "quantity": 25
        });
        let product = parser.parse_product_json(&data).unwrap();
        assert_eq!(product.stock_level, Some(25));

        // Case 4: No stock info
        let data = json!({
            "id": "126",
            "title": "Test Product 4",
            "price": 10.0
        });
        let product = parser.parse_product_json(&data).unwrap();
        assert_eq!(product.stock_level, None);
    }
}

File: backend/scraper/tiktok/parser.py
"""
TikTok Data Parser
Extract product data from TikTok Shop pages
"""

import re
import uuid
from datetime import datetime, timezone
from typing import List, Optional

from playwright.async_api import Page


class TikTokParser:
    """Parse TikTok Shop HTML/JSON to extract product data"""
    
    async def parse_product_list(self, page: Page) -> List[dict]:
        """Extract products from a product listing page"""
        products = []
        
        try:
            # Try to get data from window.__INITIAL_STATE__
            # Added try-catch block inside evaluate to prevent crashes if window object is restricted
            initial_state = await page.evaluate("""
                () => {
                    try {
                        if (window.__INITIAL_STATE__) {
                            return JSON.stringify(window.__INITIAL_STATE__);
                        }
                        // Try alternative state variables often used by frameworks
                        if (window.__NEXT_DATA__) return JSON.stringify(window.__NEXT_DATA__);
                        if (window.__NUXT__) return JSON.stringify(window.__NUXT__);
                    } catch (e) { return null; }
                    return null;
                }
            """)
            
            if initial_state:
                import json
                data = json.loads(initial_state)
                products = self._extract_from_initial_state(data)
            
            # Fallback to DOM extraction if state extraction failed or returned empty
            if not products:
                products = await self._extract_from_dom(page)
            
            return products
            
        except Exception as e:
            print(f"[Parser] Error parsing products: {e}")
            # Last resort fallback
            try:
                return await self._extract_from_dom(page)
            except Exception:
                return []
    
    async def parse_product_detail(self, page: Page) -> Optional[dict]:
        """Extract detailed product information"""
        try:
            # Try window state first
            product_data = await page.evaluate("""
                () => {
                    if (window.__INITIAL_STATE__?.product) {
                        return JSON.stringify(window.__INITIAL_STATE__.product);
                    }
                    return null;
                }
            """)
            
            if product_data:
                import json
                data = json.loads(product_data)
                return self._parse_product_json(data)
            
            # Fallback to DOM
            return await self._extract_detail_from_dom(page)
            
        except Exception as e:
            print(f"[Parser] Error parsing product detail: {e}")
            return None
    
    def _extract_from_initial_state(self, data: dict) -> List[dict]:
        """Extract products from __INITIAL_STATE__"""
        products = []
        
        # Navigate through possible data structures
        product_list = (
            data.get("products", []) or
            data.get("productList", {}).get("products", []) or
            data.get("search", {}).get("products", []) or
            []
        )
        
        for item in product_list:
            try:
                product = self._parse_product_json(item)
                if product:
                    products.append(product)
            except Exception:
                continue
        
        return products
    
    def _parse_product_json(self, data: dict) -> Optional[dict]:
        """Parse a single product from JSON data"""
        try:
            tiktok_id = str(data.get("id") or data.get("productId") or data.get("product_id", ""))
            
            if not tiktok_id:
                return None
            
            # Price handling
            price_data = data.get("price", {})
            if isinstance(price_data, dict):
                price = float(price_data.get("value", 0) or price_data.get("min", 0) or 0)
                original_price = float(price_data.get("original", 0) or price_data.get("max", 0) or 0)
                currency = price_data.get("currency", "BRL")
            else:
                price = float(price_data or 0)
                original_price = float(data.get("originalPrice", 0) or 0)
                currency = "BRL"
            
            # Images
            images = []
            image_data = data.get("images", []) or data.get("image", [])
            if isinstance(image_data, list):
                images = [img.get("url", img) if isinstance(img, dict) else img for img in image_data]
            elif isinstance(image_data, str):
                images = [image_data]
            
            image_url = images[0] if images else data.get("imageUrl") or data.get("cover", {}).get("url")
            
            # Sales data
            sales_count = self._parse_sales(data.get("salesCount") or data.get("sold", 0))
            sales_7d = self._parse_sales(data.get("sales7d", 0))
            sales_30d = self._parse_sales(data.get("sales30d", 0))
            
            # Rating
            rating_data = data.get("rating", {})
            if isinstance(rating_data, dict):
                product_rating = float(rating_data.get("average", 0) or 0)
                reviews_count = int(rating_data.get("count", 0) or 0)
            else:
                product_rating = float(rating_data or 0)
                reviews_count = int(data.get("reviewCount", 0) or 0)
            
            # Seller info
            seller = data.get("seller", {}) or data.get("shop", {})
            seller_name = seller.get("name") or seller.get("shopName")
            seller_rating = float(seller.get("rating", 0) or 0)
            
            return {
                "id": str(uuid.uuid4()),
                "tiktok_id": tiktok_id,
                "title": data.get("title") or data.get("name", ""),
                "description": data.get("description") or data.get("desc"),
                "price": price,
                "original_price": original_price if original_price > price else None,
                "currency": currency,
                "category": data.get("category") or data.get("categoryName"),
                "subcategory": data.get("subcategory") or data.get("subCategoryName"),
                "seller_name": seller_name,
                "seller_rating": seller_rating if seller_rating > 0 else None,
                "product_rating": product_rating if product_rating > 0 else None,
                "reviews_count": reviews_count,
                "sales_count": sales_count,
                "sales_7d": sales_7d,
                "sales_30d": sales_30d,
                "commission_rate": float(data.get("commissionRate", 0) or 0) or None,
                "image_url": image_url,
                "images": images,
                "video_url": data.get("videoUrl") or data.get("video", {}).get("url"),
                "product_url": data.get("url") or f"https://shop.tiktok.com/product/{tiktok_id}",
                "affiliate_url": data.get("affiliateUrl"),
                "has_free_shipping": bool(data.get("freeShipping") or data.get("hasFreeShipping")),
                "is_trending": bool(data.get("isTrending") or data.get("trending")),
                "is_on_sale": original_price > price if original_price else False,
                "in_stock": data.get("inStock", True) if "inStock" in data else data.get("stock", 1) > 0,
                "collected_at": datetime.now(timezone.utc).isoformat(),
            }
            
        except Exception as e:
            print(f"[Parser] Error parsing product JSON: {e}")
            return None
    
    async def _extract_from_dom(self, page: Page) -> List[dict]:
        """Extract products directly from DOM elements"""
        products = []
        
        product_cards = await page.query_selector_all("[data-e2e='product-card'], .product-card, .product-item")
        
        for card in product_cards:
            try:
                product = await self._parse_product_card(card)
                if product:
                    products.append(product)
            except Exception:
                continue
        
        return products
    
    async def _parse_product_card(self, card) -> Optional[dict]:
        """Parse a single product card element"""
        try:
            # Extract basic info
            title_el = await card.query_selector("[data-e2e='product-title'], .product-title, h3, h4")
            title = await title_el.inner_text() if title_el else ""
            
            price_el = await card.query_selector("[data-e2e='product-price'], .product-price, .price")
            price_text = await price_el.inner_text() if price_el else "0"
            price = self._parse_price(price_text)
            
            image_el = await card.query_selector("img")
            image_url = await image_el.get_attribute("src") if image_el else None
            
            link_el = await card.query_selector("a")
            product_url = await link_el.get_attribute("href") if link_el else None
            
            # Try to extract ID from URL
            tiktok_id = ""
            if product_url:
                match = re.search(r'/product/(\d+)', product_url)
                if match:
                    tiktok_id = match.group(1)
            
            if not tiktok_id:
                tiktok_id = str(uuid.uuid4())[:8]
            
            # Sales/rating if available
            sales_el = await card.query_selector("[data-e2e='product-sales'], .sales-count")
            sales_text = await sales_el.inner_text() if sales_el else "0"
            sales_count = self._parse_sales(sales_text)
            
            rating_el = await card.query_selector("[data-e2e='product-rating'], .rating")
            rating_text = await rating_el.inner_text() if rating_el else "0"
            rating = self._parse_rating(rating_text)
            
            return {
                "id": str(uuid.uuid4()),
                "tiktok_id": tiktok_id,
                "title": title.strip(),
                "description": None,
                "price": price,
                "original_price": None,
                "currency": "BRL",
                "category": None,
                "subcategory": None,
                "seller_name": None,
                "seller_rating": None,
                "product_rating": rating,
                "reviews_count": 0,
                "sales_count": sales_count,
                "sales_7d": 0,
                "sales_30d": 0,
                "commission_rate": None,
                "image_url": image_url,
                "images": [image_url] if image_url else [],
                "video_url": None,
                "product_url": product_url or "",
                "affiliate_url": None,
                "has_free_shipping": False,
                "is_trending": False,
                "is_on_sale": False,
                "in_stock": True,
                "collected_at": datetime.now(timezone.utc).isoformat(),
            }
            
        except Exception as e:
            print(f"[Parser] Error parsing product card: {e}")
            return None
    
    async def _extract_detail_from_dom(self, page: Page) -> Optional[dict]:
        """Extract product detail from DOM"""
        try:
            title_el = await page.query_selector("[data-e2e='product-title'], h1")
            title = await title_el.inner_text() if title_el else ""
            
            desc_el = await page.query_selector("[data-e2e='product-description'], .description")
            description = await desc_el.inner_text() if desc_el else None
            
            price_el = await page.query_selector("[data-e2e='product-price'], .price")
            price_text = await price_el.inner_text() if price_el else "0"
            price = self._parse_price(price_text)
            
            # Get all images
            images = []
            image_els = await page.query_selector_all("[data-e2e='product-image'] img, .product-images img")
            for img in image_els:
                src = await img.get_attribute("src")
                if src:
                    images.append(src)
            
            # Get product ID from URL
            url = page.url
            match = re.search(r'/product/(\d+)', url)
            tiktok_id = match.group(1) if match else str(uuid.uuid4())[:8]
            
            return {
                "id": str(uuid.uuid4()),
                "tiktok_id": tiktok_id,
                "title": title.strip(),
                "description": description,
                "price": price,
                "original_price": None,
                "currency": "BRL",
                "category": None,
                "subcategory": None,
                "seller_name": None,
                "seller_rating": None,
                "product_rating": None,
                "reviews_count": 0,
                "sales_count": 0,
                "sales_7d": 0,
                "sales_30d": 0,
                "commission_rate": None,
                "image_url": images[0] if images else None,
                "images": images,
                "video_url": None,
                "product_url": url,
                "affiliate_url": None,
                "has_free_shipping": False,
                "is_trending": False,
                "is_on_sale": False,
                "in_stock": True,
                "collected_at": datetime.now(timezone.utc).isoformat(),
            }
            
        except Exception as e:
            print(f"[Parser] Error extracting detail from DOM: {e}")
            return None
    
    def _parse_price(self, text: str) -> float:
        """Parse price from text"""
        if not text:
            return 0.0
        
        # Remove currency symbols and text
        cleaned = re.sub(r'[R$\s]', '', str(text))
        cleaned = cleaned.replace('.', '').replace(',', '.')
        
        try:
            return float(cleaned)
        except ValueError:
            return 0.0
    
    def _parse_sales(self, value) -> int:
        """Parse sales count from various formats"""
        if isinstance(value, int):
            return value
        
        if not value:
            return 0
        
        text = str(value).lower().strip()
        
        # Handle "1.2k", "3.5M" formats
        multipliers = {'k': 1000, 'm': 1000000}
        
        for suffix, mult in multipliers.items():
            if text.endswith(suffix):
                try:
                    return int(float(text[:-1]) * mult)
                except ValueError:
                    pass
        
        # Remove non-numeric
        cleaned = re.sub(r'[^\d]', '', text)
        
        try:
            return int(cleaned)
        except ValueError:
            return 0
    
    def _parse_rating(self, text: str) -> Optional[float]:
        """Parse rating from text"""
        if not text:
            return None
        
        match = re.search(r'(\d+\.?\d*)', str(text))
        if match:
            try:
                rating = float(match.group(1))
                return rating if 0 <= rating <= 5 else None
            except ValueError:
                pass
        
        return None
